pkgs/development/tools/ocaml/merlin/default.nix

```diff
@@ -1,17 +1,22 @@
-{ lib, fetchurl, buildDunePackage, yojson }:
+{ lib, fetchurl, buildDunePackage, substituteAll
+, dot-merlin-reader, dune_2, yojson, csexp, result }:
 
 buildDunePackage rec {
   pname = "merlin";
-  version = "3.3.9";
 
-  minimumOCamlVersion = "4.02.1";
+  inherit (dot-merlin-reader) src version;
 
-  src = fetchurl {
-    url = "https://github.com/ocaml/merlin/releases/download/v${version}/merlin-v${version}.tbz";
-    sha256 = "00ng8299l5rzpak8ljxzr6dgxw6z52ivm91159ahv09xk4d0y5x3";
-  };
+  minimumOCamlVersion = "4.02.3";
+
+  patches = [
+    (substituteAll {
+      src = ./fix-paths.patch;
+      dot_merlin_reader = "${dot-merlin-reader}/bin/dot-merlin-reader";
+      dune = "${dune_2}/bin/dune";
+    })
```

Comment:
What’s the rationale for this patch? IIUC, `merlin` uses either `dune` or `dot-merlin-reader` depending on the development environment: it should be able to find them in the PATH when needed.

Reply:
This ensures that `merlin` is able to find `dune` or `dot-merlin-reader` even if they are not on the PATH (e.g. if the user only has `merlin` in his `systemPackages`).

Both `dot-merlin-merlin` and `dune` are runtime dependencies of `merlin`.  As far as I understand runtime dependencies of executable needs [special handling](https://discourse.nixos.org/t/runtime-dependency-on-an-executable/1759) in Nix, either by patching the source or wrapping the executable.

Reply:
Indeed, but each user only needs one of them: if you use `dune` rather than `.merlin` files, do you want to afford `dot-merlin-reader` in your PATH?

Reply:
> but each user only needs one of them:

That's true on a project level, I don't think that's a decision we can made at build time. Right now the [logic](https://github.com/ocaml/merlin/blob/7a45a1f99ea88e68b35ad67aa67d4455caa50523/src/kernel/mconfig_dot.ml#L209-L216) is roughly: if `.merlin` files are present, use `dot-merlin-reader`, otherwise use `dune`. So one project may need `dot-merlin-reader` on the PATH, while another needs `dune`.

I definitely think we should patch the path of `dot-merlin-reader`, because that's not something that a user will likely have on his PATH and the vast majority of projects still use `.merlin` files. For `dune` it is debatable, you can argue if you are building a Dune project, you should already have `dune` on your PATH.

Reply:
Thanks for your detailed and convincing answer.